calico/felix/test/test_fetcd_api.py

```diff
@@ -0,0 +1,185 @@
+# limitations under the License.
+"""
+felix.test.test_etcd_api
+~~~~~~~~~~~
+
+Top level tests for EtcdAPI.
+"""
+import json
+import logging
+import socket
+import gevent
+from etcd import EtcdException
+from mock import Mock, call, patch
+from calico.datamodel_v1 import key_for_status, key_for_uptime
+from calico.felix.fetcd import EtcdAPI
+from calico.felix.test.base import BaseTestCase
+
+
+_log = logging.getLogger(__name__)
+
+
+class TestEtcdAPI(BaseTestCase):
+    def setUp(self):
+        super(TestEtcdAPI, self).setUp()
+        self.m_config = Mock()
+        self.m_config.IFACE_PREFIX = "tap"
+        self.m_config.ETCD_ADDR = "localhost:4001"
+        self.m_config.HOSTNAME = socket.gethostname()
+        self.m_config.RESYNC_INTERVAL = 0
+        self.m_config.REPORTING_INTERVAL_SECS = 0
+        self.m_config.REPORTING_TTL_SECS = 0
+
+    @patch('calico.felix.fetcd.EtcdAPI.write_to_etcd')
+    @patch('calico.felix.fetcd._EtcdWatcher')
+    @patch('calico.felix.fetcd.etcd')
+    def finish_setup(self, m_etcd, m_EtcdWatcher, m_write_to_etcd, **kwargs):
+        # Set configuration attributes and start etcd_api
+        for key, value in kwargs.iteritems():
+            setattr(self.m_config, key, value)
+
+        self.m_hosts_ipset = Mock()
+        self.etcd_api = EtcdAPI(self.m_config, self.m_hosts_ipset)
+
+        self.etcd_api.write_to_etcd = Mock()
+
+    def run_actor_loop(self):
+        self.etcd_api._step()
+
+    def test_write_to_etcd_actor_message(self):
+        """
+        Test write_to_etcd actor message calls client.write
+        """
+        self.m_hosts_ipset = Mock()
+        self.etcd_api = EtcdAPI(self.m_config, self.m_hosts_ipset)
+
+        with patch.object(self.etcd_api.client, 'write'):
+            self.etcd_api.write_to_etcd('key', 'value', async=True)
+            self.run_actor_loop()
+            self.assertTrue(self.etcd_api.client.write.called)
+
+    def test_update_felix_status_disabled(self):
+        """
+        Test reporting is disabled for reporting interval 0
+        """
+        self.finish_setup()
+        # We call gevent.sleep so that _status_reporting thread is executed -
+        # since no interval and ttl is given, _status_reporting thread returns
+        # instead of writing to etcd.
+        gevent.sleep(1)
+        self.assertFalse(self.etcd_api.write_to_etcd.called)
+
+    def test_update_felix_status_single(self):
+        """
+        Test felix status is updated
+        """
+        self.finish_setup(REPORTING_INTERVAL_SECS=15,
+                          REPORTING_TTL_SECS=37)
+
+        hostname = self.etcd_api._config.HOSTNAME
+        status_key = key_for_status(hostname)
+        uptime_key = key_for_uptime(hostname)
+        ttl = self.etcd_api._config.REPORTING_TTL_SECS
+
+        # We call gevent.sleep so that _status_reporting thread is executed
+        gevent.sleep(1)
+
+        status_call = call(status_key, TestIfStatus(), async=True)

```

Comment:
Can't decide if I like TestIfStatus() or not, I'd usually do this to unpack the arguments, which is fewer lines of code, but they are more ugly:
```
fn_name, args, kwargs = self.etcd_api.write_to_etcd.mock_calls[0]
status_key, if_status = args
self.assertEqual(status_key, ...)
```